build/src/lib.rs

```diff
@@ -40,50 +83,146 @@ impl Binary {
-        }
-    }
-
     // ... otherwise, request the data from the URL ...
     let bytes = reqwest::blocking::get(&format!(
         "https://github.com/bminixhofer/nlprule/releases/download/{}/{}.gz",
         version, filename
     ))?
-    .error_for_status()?
+    .error_for_status()
+    .map_err(|e| {
+        if let Some(404) = e.status().map(|x| x.as_u16()) {
+            Error::BinariesNotFound
+        } else {
+            e.into()
+        }
+    })?
     .bytes()?;
 
     let mut gz = GzDecoder::new(&bytes[..]);
     let mut buffer = Vec::new();
     gz.read_to_end(&mut buffer)?;
 
-    // ... and then cache the data at the provided file, if one was found
-    if let Some(path) = &cache_path {
-        fs::create_dir_all(path.parent().expect("path must have parent"))?;
-        fs::write(path, &buffer)?;
+    Ok(Cursor::new(buffer))
+}
+
+fn construct_cache_path(
+    version: &str,
+    lang_code: &str,
+    binary: Binary,
+    cache_dir: Option<&PathBuf>,
+    transform_path_fn: Option<&dyn TransformPathFn>,
+) -> Result<Option<PathBuf>> {
+    let filename = binary.filename(lang_code);
+
+    cache_dir
+        .map(move |dir| {
+            let path = dir.join(version).join(lang_code).join(&filename);
+            Ok(if let Some(transform_path_fn) = transform_path_fn {
+                transform_path_fn(path).map_err(Error::TransformError)?
+            } else {
+                path
+            })
+        })
+        .transpose()
+}
+
+/// Returns a `dyn Read` type, that is either directly feed
+/// from an in-memory slice or from the on disk cache.
+/// If the on-disk cache is disabled or is not present,
+/// it will attempt to download it via [`obtain_binary_from_github_release`].
+/// The data is then written into `dest`.
+/// Also updates the cache.
+fn obtain_binary_cache_or_github(
+    version: &str,
+    lang_code: &str,
+    binary: Binary,
+    cache_dir: Option<&PathBuf>,
+    transform_path_fn: Option<&dyn TransformPathFn>,
+    transform_data_fn: Option<&dyn TransformDataFn>,
+) -> Result<Box<dyn Read>> {
+    let cache_path = construct_cache_path(
+        version,
+        lang_code,
+        binary,
+        cache_dir.clone(),
+        transform_path_fn,
+    )?;
+
+    // if the file can be read, the data is already cached and the transform was applied before
+    if let Some(ref cache_path) = cache_path {
+        if let Ok(bytes) = fs::read(cache_path) {
+            return Ok(Box::new(Cursor::new(bytes)));
+        }
     }
 
-    Ok(Cursor::new(buffer))
+    // the binencoded data from github
+    let reader_binenc = obtain_binary_from_github_release(version, lang_code, binary)?;
+
+    // apply the transform if any to an intermediate buffer
+    let mut reader_transformed = if let Some(transform_data_fn) = transform_data_fn {
+        // TODO this is not optimal, the additional copy is a bit annoying
+        let mut intermediate = Box::new(Cursor::new(Vec::<u8>::with_capacity(4096 * 1024)));
```

Comment:
Let's just use `Vec::new()` here. Below (before compilation) we also allocate an empty buffer instead of using some fixed capacity.

```suggestion
        let mut intermediate = Box::new(Cursor::new(Vec::<u8>::new()));
```